File: SlapMac/SlapMac/AudioManager.swift
import Foundation
import AVFoundation
import AppKit

final class AudioManager {
    
    private var audioPlayers: [AVAudioPlayer] = []
    private var soundFiles: [URL] = []
    private var volume: Float = 1.0
    private var isEnabled = true
    
    init() {
        loadSounds()
    }
    
    var currentVolume: Float {
        get { volume }
        set {
            volume = max(0.0, min(1.0, newValue))
        }
    }
    
    var enabled: Bool {
        get { isEnabled }
        set { isEnabled = newValue }
    }
    
    var soundCount: Int {
        return soundFiles.count
    }
    
    private func loadSounds() {
        let bundle = Bundle.main
        
        // Load all supported audio files from the bundle
        let supportedExtensions = ["mp3", "wav", "aiff", "m4a", "caf", "aac"]
        
        for ext in supportedExtensions {
            if let urls = bundle.urls(forResourcesWithExtension: ext, subdirectory: nil) {
                soundFiles.append(contentsOf: urls)
            }
        }
        
        // Also check Resources subdirectory
        for ext in supportedExtensions {
            if let urls = bundle.urls(forResourcesWithExtension: ext, subdirectory: "Resources") {
                soundFiles.append(contentsOf: urls)
            }
        }
        
        if soundFiles.isEmpty {
            NSLog("[SlapMac] Warning: No audio files found in bundle")
        } else {
            NSLog("[SlapMac] Loaded \(soundFiles.count) sound(s)")
        }
        
        // Pre-load audio players for better responsiveness
        preloadPlayers()
    }
    
    private func preloadPlayers() {
        audioPlayers.removeAll()
        
        for url in soundFiles {
            do {
                let player = try AVAudioPlayer(contentsOf: url)
                player.prepareToPlay()
                player.volume = volume
                audioPlayers.append(player)
            } catch {
                NSLog("[SlapMac] Failed to load audio: \(url.lastPathComponent) - \(error.localizedDescription)")
            }
        }
    }
    
    func playRandomSound() {
        guard isEnabled, !audioPlayers.isEmpty else { return }
        
        let index = Int.random(in: 0..<audioPlayers.count)
        let player = audioPlayers[index]
        
        player.volume = volume
        player.currentTime = 0
        player.play()
        
        NSLog("[SlapMac] Playing: \(soundFiles[index].lastPathComponent)")
    }
    
    func addCustomSoundFromURL(_ url: URL) -> Bool {
        do {
            // Sanitize filename - strip path traversal characters
            let rawName = url.lastPathComponent
            let sanitized = rawName
                .replacingOccurrences(of: "..", with: "")
                .replacingOccurrences(of: "/", with: "_")
                .replacingOccurrences(of: "\\", with: "_")
            guard !sanitized.isEmpty else {
                NSLog("[SlapMac] Invalid filename")
                return false
            }
            
            // Validate it's actually an audio file
            let allowedExtensions = Set(["mp3", "wav", "aiff", "m4a", "caf", "aac"])
            guard allowedExtensions.contains(url.pathExtension.lowercased()) else {
                NSLog("[SlapMac] Unsupported audio format: \(url.pathExtension)")
                return false
            }
            
            // Copy to app support directory
            let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            let slapMacDir = appSupport.appendingPathComponent("SlapMac/Sounds", isDirectory: true)
            
            try FileManager.default.createDirectory(at: slapMacDir, withIntermediateDirectories: true)
            
            let destURL = slapMacDir.appendingPathComponent(sanitized)
            
            if FileManager.default.fileExists(atPath: destURL.path) {
                try FileManager.default.removeItem(at: destURL)
            }
            
            try FileManager.default.copyItem(at: url, to: destURL)
            
            let player = try AVAudioPlayer(contentsOf: destURL)
            player.prepareToPlay()
            player.volume = volume
            
            soundFiles.append(destURL)
            audioPlayers.append(player)
            
            NSLog("[SlapMac] Added custom sound: \(url.lastPathComponent)")
            return true
        } catch {
            NSLog("[SlapMac] Failed to add custom sound: \(error.localizedDescription)")
            return false
        }
    }
}

File: SlapMac-Android/app/src/main/kotlin/com/slapmac/app/AudioManager.kt
package com.slapmac.app

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.media.MediaPlayer
import android.util.Log

class AudioManager(private val context: Context) {
    private val soundFiles = mutableListOf<String>()
    private var mediaPlayer: MediaPlayer? = null
    var volume: Float = 1.0f
    var isEnabled = true
    val soundCount: Int get() = soundFiles.size

    init {
        loadSounds()
    }

    private fun loadSounds() {
        val extensions = setOf("mp3", "wav", "m4a", "ogg")
        try {
            val assets = context.assets.list("audio") ?: emptyArray()
            for (file in assets) {
                val ext = file.substringAfterLast('.', "").lowercase()
                if (ext in extensions) {
                    soundFiles.add("audio/$file")
                }
            }
            Log.d("SlapMac", "Loaded ${soundFiles.size} sound(s)")
        } catch (e: Exception) {
            Log.e("SlapMac", "Failed to load sounds: ${e.message}")
        }
    }

    fun playRandomSound() {
        if (!isEnabled || soundFiles.isEmpty()) return

        try {
            mediaPlayer?.release()
            val file = soundFiles.random()
            val afd: AssetFileDescriptor = context.assets.openFd(file)
            mediaPlayer = MediaPlayer().apply {
                setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
                afd.close()
                setVolume(volume.coerceIn(0f, 1f), volume.coerceIn(0f, 1f))
                prepare()
                start()
            }
            Log.d("SlapMac", "Playing: $file")
        } catch (e: Exception) {
            Log.e("SlapMac", "Playback error: ${e.message}")
        }
    }

    fun release() {
        mediaPlayer?.release()
        mediaPlayer = null
    }
}
